modules/currency-exchange/src/lib.rs

```diff
@@ -0,0 +1,414 @@
+	/// Peer blockchain type.
+	type PeerBlockchain: Blockchain;
+	/// Peer blockchain transaction parser.
+	type PeerMaybeLockFundsTransaction: MaybeLockFundsTransaction<
+		Transaction = <Self::PeerBlockchain as Blockchain>::Transaction,
+	>;
+	/// Map between blockchains recipients.
+	type RecipientsMap: RecipientsMap<
+		PeerRecipient = <Self::PeerMaybeLockFundsTransaction as MaybeLockFundsTransaction>::Recipient,
+		Recipient = Self::AccountId,
+	>;
+	/// This blockchain currency amount type.
+	type Amount;
+	/// Converter from peer blockchain currency type into current blockchain currency type.
+	type CurrencyConverter: CurrencyConverter<
+		SourceAmount = <Self::PeerMaybeLockFundsTransaction as MaybeLockFundsTransaction>::Amount,
+		TargetAmount = Self::Amount,
+	>;
+	/// Something that could grant money.
+	type DepositInto: DepositInto<Recipient = Self::AccountId, Amount = Self::Amount>;
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Invalid peer blockchain transaction provided.
+		InvalidTransaction,
+		/// Peer transaction has invalid amount.
+		InvalidAmount,
+		/// Peer transaction has invalid recipient.
+		InvalidRecipient,
+		/// Cannot map from peer recipient to this blockchain recipient.
+		FailedToMapRecipients,
+		/// Failed to convert from peer blockchain currency to this blockhain currency.
+		FailedToConvertCurrency,
+		/// Deposit has failed.
+		DepositFailed,
+		/// Transaction is not finalized.
+		UnfinalizedTransaction,
+		/// Transaction funds are already claimed.
+		AlreadyClaimed,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Imports lock fund transaction of the peer blockchain.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn import_peer_transaction(
+			origin,
+			proof: <<T as Trait>::PeerBlockchain as Blockchain>::TransactionInclusionProof,
+		) -> DispatchResult {
+			let submitter = frame_system::ensure_signed(origin)?;
+
+			// ensure that transaction is included in finalized block that we know of
+			let transaction = <T as Trait>::PeerBlockchain::verify_transaction_inclusion_proof(
+				&proof,
+			).ok_or_else(|| Error::<T>::UnfinalizedTransaction)?;
+
+			// parse transaction
+			let transaction = <T as Trait>::PeerMaybeLockFundsTransaction::parse(&transaction)
+				.map_err(Error::<T>::from)?;
+			let transfer_id = transaction.id;
+			ensure!(
+				!Transfers::<T>::contains_key(&transfer_id),
+				Error::<T>::AlreadyClaimed
+			);
+
+			// grant recipient
+			let recipient = T::RecipientsMap::map(transaction.recipient).map_err(Error::<T>::from)?;
+			let amount = T::CurrencyConverter::convert(transaction.amount).map_err(Error::<T>::from)?;
+			T::DepositInto::deposit_into(recipient, amount).map_err(Error::<T>::from)?;
+
+			// remember that we have accepted this transfer
+			Transfers::<T>::insert(transfer_id, ());
```

Comment:
Should this come before the `deposit_into` step? I'm not sure how this would happen, but I'm imagining a scenario in which we deposit funds and then fail to reach here, which would lead to us being able to claim stuff multiple times

Reply:
That's a good point - I'm also not sure how this could happen. The only case I could think of, should be impossible in Substrate - i.e. if implementation of `RecipientsMap`, `CurrencyConverter` or `DepositInto` will be calling into some untrusted (contract?) code and this code will be able to call back into 'system code' (here: submit same proof again). But if contract is able to call system code, then it also will be able to change any storage value (balance, ...). So this change won't change anything.

So the proper fix imo, is not to allow untrusted code to call into system code at all. Afaik it's already how it works. However, if you still think it'd be better to protect against that, I could change code to something like:
```
Transfers::insert();
if let Err(error) = deposit_into() {
  Transfers::remove();
  return Err(error);
}
```

It is something against regular "updating state after all checks are completed" pattern, though.

Reply:
I took a look at the code again and I still don't know how it might happen, so no need to change the code to that. It wouldn't hurt for @tomusdrw should take a look and see if there's anything we missed

Reply:
Yeah, I was thinking about this when reviewing earlier, but I think the code is correct. To make it more future proof I'd suggest to group the two statements that are related into one block, like so:
```rust
let recipient = T::RecipientsMap::map(transaction.recipient).map_err(Error::<T>::from)?;
let amount = T::CurrencyConverter::convert(transaction.amount).map_err(Error::<T>::from)?;

// make sure to update the mapping if we deposit successfully to avoid double spending,
// i.e. whenever `deposit_into` is successful we MUST update `Transfers`.
{
    T::DepositInto::deposit_into(recipient, amount).map_err(Error::<T>::from)?;
    Transfers::<T>::insert(transfer_id, ())
}
```

This will make it harder to introduce `?` statements in between `deposit_into` and `Transfers::insert`.